src/interface/misc.rs

```diff
@@ -363,3 +365,86 @@ impl std::fmt::Debug for RawCondvar {
         f.write_str("<condvar>")
     }
 }
+
+/*
+* RustSemaphore is the rust version of sem_t
+*/
+#[derive(Debug)]
+pub struct RustSemaphore {
+    pub value: RustAtomicU32,
+    pub is_shared: RustAtomicBool,
+}
+
+// Semaphore implementation
+// we busy wait on lock if value is 0, otherwise we decrease the value
+// unlock will increase value up to SEM_VALUE_MAX
+impl RustSemaphore {
+    pub fn new(value_handle: u32, is_shared: bool) -> Self {
+        Self {
+            value: RustAtomicU32::new(value_handle),
+            is_shared: RustAtomicBool::new(is_shared),
+        }
+    }
+
+    pub fn lock(&self) -> bool{
+        while self.value.load(RustAtomicOrdering::Relaxed) == 0 { interface::lind_yield(); }
+
+        let result = self.value.fetch_update(RustAtomicOrdering::Relaxed, RustAtomicOrdering::Relaxed, |x| {
```

Comment:
This general code structure to update the semaphore is used in almost all the new functions here so maybe just use a helper function, like `update_semaphore` to get rid of duplicated code.